# Genes under strong positive selection in the ENIGMA environment
* The Strainfinder v1 app was applied to 6 ENIGMA MAGs as a demonstration.
* We identified and visualized the strain genotypes of 6 MAGs in ENIGMA metagenome samples.
* We further screened for genes under strong positive selection in the ENIGMA environment by the criteria of multiple point mutations and high nonsynonymous mutation ratio (dN/dS > 1).
* We found that genes responsible for carbohydrate metabolism and transport, regulatory factors, and antibiotic resistance were the main targets of positive selection.

![Antibiotic_resistance_genes.pdf](https://raw.githubusercontent.com/caozhichongchong/meta_decoder/plots/Antibiotic_resistance_genes.pdf)
Fig 1. Antibiotic resistance genes under strong positive selection

![Carbonhydrate_metabolism.pdf](https://raw.githubusercontent.com/caozhichongchong/meta_decoder/plots/Carbonhydrate_metabolism.pdf)
Fig 2. Genes responsive for carbonhydrate metabolism under strong positive selection

![Regulation.pdf](https://raw.githubusercontent.com/caozhichongchong/meta_decoder/plots/Regulation.pdf)
Fig 3. Regulatory factors under strong positive selection
